lib/spack/spack/bootstrap.py

```diff
@@ -97,48 +99,139 @@ def _try_import_from_store(module, abstract_spec_str):
         if str(bincache_platform) == 'cray':
             bincache_platform = spack.platforms.linux.Linux()
             with spack.architecture.use_platform(bincache_platform):
-                abstract_spec = spack.spec.Spec(
-                    abstract_spec_str + ' ^' + spec_for_current_python()
-                )
+                abstract_spec = spack.spec.Spec(abstract_spec_str)
+        return abstract_spec, bincache_platform
 
-        # Read information on verified clingo binaries
-        json_filename = '{0}.json'.format(module)
+    def _read_metadata(self, package_name):
+        """Return metadata about the given package."""
+        json_filename = '{0}.json'.format(package_name)
         json_path = os.path.join(
             spack.paths.share_path, 'bootstrap', self.name, json_filename
         )
         with open(json_path) as f:
             data = json.load(f)
-
-        buildcache = spack.main.SpackCommand('buildcache')
-        # Ensure we see only the buildcache being used to bootstrap
-        mirror_scope = spack.config.InternalConfigScope(
+        return data
+
+    def _install_by_hash(self, pkg_hash, pkg_sha256, index, bincache_platform):
+        global _buildcache_cmd
+
+        if _buildcache_cmd is None:
+            _buildcache_cmd = spack.main.SpackCommand('buildcache')
+
+        index_spec = next(x for x in index if x.dag_hash() == pkg_hash)
+        # Reconstruct the compiler that we need to use for bootstrapping
+        compiler_entry = {
+            "modules": [],
+            "operating_system": str(index_spec.os),
+            "paths": {
+                "cc": "/dev/null",
+                "cxx": "/dev/null",
+                "f77": "/dev/null",
+                "fc": "/dev/null"
+            },
+            "spec": str(index_spec.compiler),
+            "target": str(index_spec.target.family)
+        }
+        with spack.architecture.use_platform(bincache_platform):
+            with spack.config.override(
+                    'compilers', [{'compiler': compiler_entry}]
+            ):
+                spec_str = '/' + pkg_hash
+                install_args = [
+                    'install',
+                    '--sha256', pkg_sha256,
+                    '--only-root',
+                    '-a', '-u', '-o', '-f', spec_str
+                ]
+                _buildcache_cmd(*install_args, fail_on_error=False)
+
+    @property
+    def mirror_scope(self):
+        return spack.config.InternalConfigScope(
             'bootstrap', {'mirrors:': {self.name: self.url}}
         )
-        with spack.config.override(mirror_scope):
+
+    def try_import(self, module, abstract_spec_str):
+        if _try_import_from_store(module, abstract_spec_str):
+            return True
+
+        abstract_spec, bincache_platform = self._spec_and_platform(
+            abstract_spec_str + ' ^' + spec_for_current_python()
+        )
+        data = self._read_metadata(module)
```

Comment:
This should only be necessary for `clingo` -- all other python modules we can bootstrap gpg and do normal buildcache installs for them

Reply:
Sure, I would modify that on the first PR that bootstraps another python module if you agree. I think one candidate might be #25371

Reply:
Thinking about that, for most pure python modules it wouldn't make sense to rely on a buildcache, right? For instance installing `pytest` or `flake8` would not be slower or less portable if done from sources I think 🤔

Reply:
True, it might not be worth bothering for the python modules we currently use.